opm/simulators/linalg/ParallelOverlappingILU0.hpp

```diff
@@ -1036,5 +1095,365 @@ class ParallelOverlappingILU0
+                  the vertices with the same color.
+    */
+    template<class BlockType, class Alloc>
+    GhostLastParallelOverlappingILU0 (const Dune::BCRSMatrix<BlockType,Alloc>& A,
+                                      const field_type w, MILU_VARIANT milu, unsigned interiorSize,
+                                      bool redblack=false,
+                                      bool reorder_sphere=true)
+        : GhostLastParallelOverlappingILU0( A, 0, w, milu, interiorSize, redblack, reorder_sphere )
+    {
+    }
+
+    /*! \brief Constructor.
+      Constructor gets all parameters to operate the prec.
+      \param A      The matrix to operate on.
+      \param comm   communication object, e.g. Dune::OwnerOverlapCopyCommunication
+      \param w      The relaxation factor.
+      \param milu   The modified ILU variant to use. 0 means traditional ILU. \see MILU_VARIANT.
+      \param redblack Whether to use a red-black ordering.
+      \param reorder_sphere If true, we start the reordering at a root node.
+                            The vertices on each layer aound it (same distance) are
+                            ordered consecutivly. If false, we preserver the order of
+                            the vertices with the same color.
+    */
+    template<class BlockType, class Alloc>
+    GhostLastParallelOverlappingILU0 (const Dune::BCRSMatrix<BlockType,Alloc>& A,
+                                      const ParallelInfo& comm, const field_type w,
+                                      MILU_VARIANT milu, unsigned interiorSize,
+                                      bool redblack=false,
+                                      bool reorder_sphere=true)
+        : lower_(),
+          upper_(),
+          inv_(),
+          comm_(&comm), w_(w),
+          relaxation_( std::abs( w - 1.0 ) > 1e-15 ),
+          interiorSize_(interiorSize)
+    {
+        // BlockMatrix is a Subclass of FieldMatrix that just adds
+        // methods. Therefore this cast should be safe.
+        init( reinterpret_cast<const Matrix&>(A), 0, milu, redblack,
+              reorder_sphere );
+    }
+
+    /*!
+      \brief Prepare the preconditioner.
+      \copydoc Preconditioner::pre(X&,Y&)
+    */
+    virtual void pre (Domain& x, Range& b) override
+    {
+        DUNE_UNUSED_PARAMETER(x);
+        DUNE_UNUSED_PARAMETER(b);
+    }
+
+    /*!
+      \brief Apply the preconditoner.
+      \copydoc Preconditioner::apply(X&,const Y&)
+    */
+    virtual void apply (Domain& v, const Range& d) override
+    {
+        Range& md = const_cast<Range&>(d);;
+
+        // iterator types
+        typedef typename Range ::block_type  dblock;
+        typedef typename Domain::block_type  vblock;
+
+        const size_type iEnd = lower_.rows();
+        const size_type lastRow = iEnd - 1;
+        size_type interiorStart = iEnd - interiorSize_;
+        size_type lowerLoopEnd = interiorSize_;
+        if( iEnd != upper_.rows() )
```

Comment:
This looks like code duplication. Where is the original code coming from? Could we factor it (the loops) out to function that get an index or iterator as the end and just use the correct one here and the default at other places.

Reply:
I believe this was resolved by removing the GhostLastParallaelOverlappingILU0 class.